packages/core/strapi/lib/types/core/common/controller.d.ts

```diff
@@ -0,0 +1,8 @@
+import type koa from 'koa';
+
+export type ControllerHandler = (
+  context: koa.ExtendableContext,
+  next: koa.Next
+) => Promise<void> | void | Promise<unknown> | unknown;
```

Comment:
Just a thought - maybe you already thought, wouldn't it be nice to be able to type your return as a generic? Default would be unknown if you don't pass one and you could still make this union?

Reply:
This makes me realize that I originally wanted to use `Promise<void> | void` only :thinking: 
But that's not that good either since it's technically not true & even if we don't want to encourage that, nothing is preventing you from actually returning something.

Reply:
Let's add a generic for now :man_shrugging:

Reply:
Also, (last message of my monologue, I promise) controllers & services types are only defined because I needed them to define their respective registries types, but they will likely be modified in the future

Reply:
[Update the controller type to make it more generic](https://github.com/strapi/strapi/pull/16724/commits/8f1f784cc7d6a29004e242917d7b7d884aed6703)

Reply:
> but they will likely be modified in the future

don't forget #agile

Reply:
![image](https://github.com/strapi/strapi/assets/25851739/aee96bd5-816f-4241-a353-68607431b605)